Update HPLT to 2.0

eu9ene opened the pull request:
- use hplt 2.0 https://hplt-project.org/datasets/v2.0
- filter by document score instead of segment fluency (there's no segment fluency score in hplt 2.0)
- add statistics about filtering
- add a separate parameter to enable merging to preserve length filtering
- update tests and extract them to a separate file
- refactor the downloader to a class so that it's easier to understand

closes #884 
closes #944

ZJaume commented on pipeline/data/download-mono.py:
```diff
@@ -57,10 +54,10 @@ def main(args_list: Optional[list[str]] = None) -> None:
     )
     parser.add_argument(
-        "--hlpt_min_fluency",
+        "--hlpt_min_doc_score",
```
Just a minor thing. Maybe this PR could also fix the HPLT names? It is mistakenly written HLPT in here and in other parts of the code.

ZJaume reviewed (approved):
Maybe it is useful, but I don't know where it should mentioned. Each HPLT dataset has its own analytics report linked in the website: https://hplt-project.org/datasets/v2.0. For example: https://github.com/hplt-project/data-analytics-tool/blob/main/reports/mono-2.0/HPLT-v2-fra_Latn.lite.pdf
This can help choosing thresholds, as the score distribution is shown there.

eu9ene commented:
I realized that I didn't update tests

eu9ene commented:
> > Maybe it is useful, but I don't know where it should mentioned. Each HPLT dataset has its own analytics report linked in the website: https://hplt-project.org/datasets/v2.0. For example: https://github.com/hplt-project/data-analytics-tool/blob/main/reports/mono-2.0/HPLT-v2-fra_Latn.lite.pdf This can help choosing thresholds, as the score distribution is shown there.
> 
> Yeah, it's a good idea to show the link somewhere. I looked at those for document scores. Such a report is missing for zh_Hans though.

Actually, it's there, the link is missing on the web site. I think it can be automated based on the yaml files: https://github.com/hplt-project/data-analytics-tool/blob/main/reports/mono-2.0/HPLT-v2-zho_Hans.lite.yaml
At least we could check that, say, not less than 10% of data fits the threshold criteria. I'll file a separate issue about this.

eu9ene commented:
Ok, while I was looking at tests I also added more statistics about filtering and refactored the downloader a bit. @gregtatum please have a look when you have a chance.

eu9ene commented on pipeline/data/importers/mono/hplt.py:
```diff
@@ -70,33 +72,56 @@ def __init__(self, dataset_path: Path) -> None:
+    """
+    locale = icu.Locale(lang_iso6931)
+    # add default script
+    locale = icu.Locale.addLikelySubtags(locale)
```
Ok, this doesn't work for Korean... It returns Kore which is a mix of Hang and Hani. I wonder if we should specify the script manually.

eu9ene replied:
https://en.wikipedia.org/wiki/ISO_15924#Kore

gregtatum commented on pipeline/data/importers/mono/hplt.py:
```diff
@@ -105,129 +133,170 @@ def load_shuffled_shard_urls(language: str) -> list[str]:
+
+            # HPLT 2.0 uses document level scores
+            if overall_doc_score < self.hplt_min_doc_score:
+                self.stats.filtered_doc_score.value += 1
```
Thanks for adding the statistics counting.

eu9ene replied:
The detailed stats json is very helpful to understand what's happening with filtering. Thank you for adding support for this! I think we should output such json files for all filtering steps we have but it will require refactoring bash to python...

gregtatum commented on tests/test_hplt.py:
```diff
@@ -0,0 +1,196 @@
+    return DataDir("test_hplt")
+
+
+def test_mono_hplt_merging(data_dir: DataDir):
```
Thanks for the run_task tests!

gregtatum reviewed (approved):
Thanks, this looks good to me. I'm not sure why the taskcluster tests are failing, but we should probably figure that out before merging.

One other thought is now that we have git LFS set up for the project it might be worth moving the `zst` files over to it. This is optional, but will help the repo's size not get bloated over time.

eu9ene commented on taskcluster/configs/config.prod.yml:
```diff
@@ -49,17 +49,18 @@ experiment:
       default-threshold: 0.0
       dataset-thresholds:
+        # We already filter it by document score, remove only the noisiest segments
+        hplt_mono_v2_0: 0.7
```
I want quality to be higher for back-translations, I'll clarify

eu9ene commented on taskcluster/configs/config.prod.yml:
```diff
@@ -73,13 +74,14 @@ experiment:
+  hplt-min-doc-score:
     # Noiser data should be fine for distillation.
-    mono-src: 0.8
+    mono-src: 7.0
```
I asked Jaume about this and he said the scores were specifically designed to mean the same across all languages, so those thresholds should be safer

eu9ene commented:
> Thanks, this looks good to me. I'm not sure why the taskcluster tests are failing, but we should probably figure that out before merging.
> 
> One other thought is now that we have git LFS set up for the project it might be worth moving the `zst` files over to it. This is optional, but will help the repo's size not get bloated over time.

It's a mystery what happened to Taskgraph tests. They pass locally. I think it has something to do with machine reimaging. Just creating a directory doesn't help, there's another error. Maybe @bhearsum has some insights.

bhearsum commented:
> > Thanks, this looks good to me. I'm not sure why the taskcluster tests are failing, but we should probably figure that out before merging.
> > One other thought is now that we have git LFS set up for the project it might be worth moving the `zst` files over to it. This is optional, but will help the repo's size not get bloated over time.
> 
> It's a mystery what happened to Taskgraph tests. They pass locally. I think it has something to do with machine reimaging. Just creating a directory doesn't help, there's another error. Maybe @bhearsum has some insights.

It looks to me like this is due to Poetry getting upgraded when the docker images were rebuilt here. (This happens because [the installations unpinned](https://github.com/mozilla/translations/blob/b08737d1dec095d281bc7d924feff06869ada4fa/taskcluster/docker/test/Dockerfile#L23) - which you might consider changing.) The [previous docker image](https://firefox-ci-tc.services.mozilla.com/tasks/BdKH0ifwRYGGWMSK6ToLvQ) had 1.8.4. [The new one](https://firefox-ci-tc.services.mozilla.com/tasks/J7bPMpf0Rg-3nNg2wXKCwg) has 2.0.1. [2.0.0](https://github.com/python-poetry/poetry/blob/main/CHANGELOG.md#200---2025-01-05) had a bunch of changes, but I suspect it is [Actually switch the directory when using --directory/-C](https://github.com/python-poetry/poetry/pull/9831) that caused the issue; it sounds like it expects whatever directory is given to exist.

eu9ene commented:
> > > Thanks, this looks good to me. I'm not sure why the taskcluster tests are failing, but we should probably figure that out before merging.
> > > One other thought is now that we have git LFS set up for the project it might be worth moving the `zst` files over to it. This is optional, but will help the repo's size not get bloated over time.
> > 
> > 
> > It's a mystery what happened to Taskgraph tests. They pass locally. I think it has something to do with machine reimaging. Just creating a directory doesn't help, there's another error. Maybe @bhearsum has some insights.
> 
> It looks to me like this is due to Poetry getting upgraded when the docker images were rebuilt here. (This happens because [the installations unpinned](https://github.com/mozilla/translations/blob/b08737d1dec095d281bc7d924feff06869ada4fa/taskcluster/docker/test/Dockerfile#L23) - which you might consider changing.) The [previous docker image](https://firefox-ci-tc.services.mozilla.com/tasks/BdKH0ifwRYGGWMSK6ToLvQ) had 1.8.4. [The new one](https://firefox-ci-tc.services.mozilla.com/tasks/J7bPMpf0Rg-3nNg2wXKCwg) has 2.0.1. [2.0.0](https://github.com/python-poetry/poetry/blob/main/CHANGELOG.md#200---2025-01-05) had a bunch of changes, but I suspect it is [Actually switch the directory when using --directory/-C](https://github.com/python-poetry/poetry/pull/9831) that caused the issue; it sounds like it expects whatever directory is given to exist.

Thank you for investigating! I pinned the dependency and it works now. We can upgrade it later separately if needed.
